File: apps/gm-console-web/scripts/render-screenshots.js
// render-screenshots.js — gm-console Web SPA screenshot pipeline (CI-grade)
//
// Purpose
// -------
// Drive a headless Chromium (via Playwright, when available) through the 7 hero
// shots documented in docs/commercial/SCREENSHOTS_BRIEF.md, at the exact pixel
// dimensions Apple/Google review templates require.
//
// Why this file is here
// ---------------------
// - This worker-E host has no Node-based browser driver installed. So the file
//   is not directly runnable on the host; the spec it ships with (SPEC.md)
//   describes the exact CI invocation.
// - It exists so future devs (and the CI workflow) have a single source of
//   truth for which URL renders which shot, at what size, into which path.
//
// Conventions
// -----------
// - All output goes under apps/gm-console-web/dist/screenshots/<device>/.
// - File names are stable: <shot-id>__<device-id>__<width>x<height>.png.
// - The dist/screenshots/INDEX.md file lists every (shot, device) tuple with a
//   NULL placeholder until the golden PNG is committed.
//
// Invocation (CI)
// ---------------
//   node apps/gm-console-web/scripts/render-screenshots.js \
//        --base-url http://127.0.0.1:8080 \
//        --out-dir  apps/gm-console-web/dist/screenshots \
//        --device-filter iphone-6.7,android-phone-portrait
//
//   # All devices if --device-filter omitted.
//   # --update-goldens writes the rendered PNGs into apps/gm-console-web/dist/screenshots/goldens/.
//
// Exit codes
// ----------
//   0   all renders succeeded and were either equal to goldens or --update-goldens was set
//   2   Playwright/Puppeteer not installed (CI installs them via npx --yes playwright)
//   3   Browser launch failed (missing system libs on the runner)
//   4   Image diff exceeded threshold (golden regression)
//
// No real PNGs are committed by this script; the deliverable is the spec +
// the INDEX catalog. PNGs land via CI artifacts.

'use strict';

const path = require('node:path');
const fs = require('node:fs');

// --- Shot catalog (mirrors docs/commercial/SCREENSHOTS_BRIEF.md) -------------
// shot-id           | route          | caption                       | device-id          | width x height
const SHOTS = [
  { id: 'login',            route: '/login',            caption: 'Sign in to your tenant',           deviceId: 'iphone-6.7',          w: 1290, h: 2796 },
  { id: 'login',            route: '/login',            caption: 'Sign in to your tenant',           deviceId: 'iphone-6.5',          w: 1242, h: 2688 },
  { id: 'login',            route: '/login',            caption: 'Sign in to your tenant',           deviceId: 'android-phone',       w: 1080, h: 1920 },
  { id: 'tenants',          route: '/tenants',          caption: 'Switch between workspaces',         deviceId: 'iphone-6.7',          w: 1290, h: 2796 },
  { id: 'pipelines',        route: '/pipelines',        caption: 'Pipelines at a glance',            deviceId: 'iphone-6.7',          w: 1290, h: 2796 },
  { id: 'pipelines',        route: '/pipelines',        caption: 'Pipelines at a glance',            deviceId: 'android-phone-portrait', w: 1080, h: 1920 },
  { id: 'pipeline-detail',  route: '/pipelines/1',      caption: 'Inspect a run, retry in one tap',  deviceId: 'iphone-6.7',          w: 1290, h: 2796 },
  { id: 'incidents',        route: '/incidents',        caption: 'Ack and resolve live alerts',      deviceId: 'iphone-6.7',          w: 1290, h: 2796 },
  { id: 'audit',            route: '/audit',            caption: 'Immutable remediation trail',      deviceId: 'iphone-6.7',          w: 1290, h: 2796 },
  { id: 'settings',         route: '/settings',         caption: 'Tokens, flavor, build info',       deviceId: 'iphone-6.7',          w: 1290, h: 2796 },
  // Optional tablet coverage (iPad 12.9") — enabled when device-filter=ipad-12.9.
  { id: 'pipelines',        route: '/pipelines',        caption: 'Pipelines at a glance',            deviceId: 'ipad-12.9',           w: 2048, h: 2732 },
];

// --- CLI parsing -------------------------------------------------------------
function parseArgs(argv) {
  const out = { baseUrl: 'http://127.0.0.1:8080', outDir: 'apps/gm-console-web/dist/screenshots', deviceFilter: null, updateGoldens: false };
  for (let i = 2; i < argv.length; i++) {
    const a = argv[i];
    if (a === '--base-url')         out.baseUrl = argv[++i];
    else if (a === '--out-dir')     out.outDir = argv[++i];
    else if (a === '--device-filter') out.deviceFilter = argv[++i].split(',').map(s => s.trim()).filter(Boolean);
    else if (a === '--update-goldens') out.updateGoldens = true;
  }
  return out;
}

// --- Driver loader (tries playwright, falls back to puppeteer) ---------------
async function loadDriver() {
  // Lazy-require so missing modules surface as a friendly error code (2).
  try {
    const playwright = require('playwright');
    return { kind: 'playwright', api: playwright };
  } catch (_) { /* fall through */ }
  try {
    const puppeteer = require('puppeteer');
    return { kind: 'puppeteer', api: puppeteer };
  } catch (_) { /* fall through */ }
  const err = new Error('Neither playwright nor puppeteer is installed in this environment.');
  err.exitCode = 2;
  throw err;
}

// --- One render --------------------------------------------------------------
async function renderOne(driver, { baseUrl, outDir, shot, updateGoldens }) {
  const filename = `${shot.id}__${shot.deviceId}__${shot.w}x${shot.h}.png`;
  const targetDir = updateGoldens ? path.join(outDir, 'goldens') : path.join(outDir, shot.deviceId);
  fs.mkdirSync(targetDir, { recursive: true });
  const outPath = path.join(targetDir, filename);

  let page;
  if (driver.kind === 'playwright') {
    const browser = await driver.api.chromium.launch({ headless: true });
    const ctx = await browser.newContext({
      viewport: { width: shot.w, height: shot.h },
      deviceScaleFactor: 1,
      colorScheme: 'light',
    });
    page = await ctx.newPage();
    await page.goto(new URL(shot.route, baseUrl).toString(), { waitUntil: 'networkidle' });
    // Wait for the shell to declare readiness (see dist/index.html hydration hook).
    await page.waitForFunction(() => document.documentElement.dataset.shellReady === '1', null, { timeout: 5000 })
      .catch(() => { /* static shell may not set this; harmless */ });
    await page.screenshot({ path: outPath, fullPage: false, type: 'png' });
    await browser.close();
  } else {
    // puppeteer path
    const browser = await driver.api.launch({ headless: 'new', args: ['--no-sandbox'] });
    page = await browser.newPage();
    await page.setViewport({ width: shot.w, height: shot.h, deviceScaleFactor: 1 });
    await page.goto(new URL(shot.route, baseUrl).toString(), { waitUntil: 'networkidle0' });
    await page.screenshot({ path: outPath, fullPage: false, type: 'png' });
    await browser.close();
  }
  return outPath;
}

// --- Image diff (lightweight, no native deps) --------------------------------
// Compares byte size and PNG signature only; sufficient to flag a regression
// when the golden exists. Real pixel-level diff is delegated to the CI step.
function quickDiff(renderedPath, goldenPath) {
  if (!fs.existsSync(goldenPath)) return { status: 'no-golden' };
  const a = fs.statSync(renderedPath).size;
  const b = fs.statSync(goldenPath).size;
  const drift = Math.abs(a - b) / Math.max(a, 1);
  return { status: drift > 0.02 ? 'drift' : 'ok', driftBytes: a - b };
}

// --- Entry point -------------------------------------------------------------
async function main() {
  const args = parseArgs(process.argv);
  const shots = args.deviceFilter
    ? SHOTS.filter(s => args.deviceFilter.includes(s.deviceId))
    : SHOTS;

  const driver = await loadDriver();
  const summary = [];
  for (const shot of shots) {
    const rendered = await renderOne(driver, { ...args, shot });
    const golden = path.join(args.outDir, 'goldens', path.basename(rendered));
    const diff = quickDiff(rendered, golden);
    summary.push({ shot, rendered, golden, diff });
    // eslint-disable-next-line no-console
    console.log(`[render] ${shot.id} @ ${shot.deviceId} (${shot.w}x${shot.h}) -> ${rendered} (${diff.status})`);
  }

  // Non-zero exit on drift so CI can fail loudly.
  const anyDrift = summary.some(s => s.diff.status === 'drift');
  process.exitCode = anyDrift ? 4 : 0;
}

main().catch((err) => {
  // eslint-disable-next-line no-console
  console.error(`[render] ${err.message}`);
  if (err.exitCode) process.exit(err.exitCode);
  process.exit(1);
});

File: apps/gm-console-web/scripts/render-screenshots.py
#!/usr/bin/env python3
"""render-screenshots.py — gm-console Web SPA screenshot pipeline (Python mirror).

This is the Python twin of render-screenshots.js. Either driver can be wired
into CI; the JavaScript file is the canonical implementation because Playwright
ships an official Node binding. The Python mirror exists so:

  * CI runners without Node can still execute the pipeline.
  * Local contributors who prefer `python` over `node` have a familiar entry.

The shot catalog, device dimensions, and golden-image conventions are kept in
sync with render-screenshots.js and docs/commercial/SCREENSHOTS_BRIEF.md.
"""

from __future__ import annotations

import argparse
import os
import sys
from dataclasses import dataclass
from pathlib import Path
from typing import Iterable


@dataclass(frozen=True)
class Shot:
    shot_id: str
    route: str
    caption: str
    device_id: str
    width: int
    height: int

    @property
    def filename(self) -> str:
        return f"{self.shot_id}__{self.device_id}__{self.width}x{self.height}.png"


# Mirror of the JavaScript catalog. Keep these two in sync.
SHOTS: tuple[Shot, ...] = (
    Shot("login",           "/login",       "Sign in to your tenant",          "iphone-6.7",            1290, 2796),
    Shot("login",           "/login",       "Sign in to your tenant",          "iphone-6.5",            1242, 2688),
    Shot("login",           "/login",       "Sign in to your tenant",          "android-phone",         1080, 1920),
    Shot("tenants",         "/tenants",     "Switch between workspaces",        "iphone-6.7",            1290, 2796),
    Shot("pipelines",       "/pipelines",   "Pipelines at a glance",           "iphone-6.7",            1290, 2796),
    Shot("pipelines",       "/pipelines",   "Pipelines at a glance",           "android-phone-portrait", 1080, 1920),
    Shot("pipeline-detail", "/pipelines/1", "Inspect a run, retry in one tap", "iphone-6.7",            1290, 2796),
    Shot("incidents",       "/incidents",   "Ack and resolve live alerts",     "iphone-6.7",            1290, 2796),
    Shot("audit",           "/audit",       "Immutable remediation trail",     "iphone-6.7",            1290, 2796),
    Shot("settings",        "/settings",    "Tokens, flavor, build info",      "iphone-6.7",            1290, 2796),
    Shot("pipelines",       "/pipelines",   "Pipelines at a glance",           "ipad-12.9",             2048, 2732),
)


def _load_driver():
    """Try Playwright (sync) first, then a manual Chromium subprocess fallback."""
    try:
        from playwright.sync_api import sync_playwright  # type: ignore
        return ("playwright", sync_playwright)
    except Exception:
        pass
    try:
        from playwright.async_api import async_playwright  # type: ignore
        return ("playwright-async", async_playwright)
    except Exception:
        pass
    raise SystemExit(2)  # No driver installed; CI installs playwright via pip.


def _render_one(driver, base_url: str, out_dir: Path, shot: Shot, update_goldens: bool) -> Path:
    target_dir = out_dir / ("goldens" if update_goldens else shot.device_id)
    target_dir.mkdir(parents=True, exist_ok=True)
    out_path = target_dir / shot.filename

    kind, api = driver
    if kind.startswith("playwright"):
        with api() as p:
            browser = p.chromium.launch(headless=True)
            ctx = browser.new_context(viewport={"width": shot.width, "height": shot.height},
                                      device_scale_factor=1, color_scheme="light")
            page = ctx.new_page()
            page.goto(f"{base_url.rstrip('/')}{shot.route}", wait_until="networkidle")
            page.screenshot(path=str(out_path), full_page=False, type="png")
            browser.close()
    else:
        raise RuntimeError(f"unsupported driver kind: {kind}")
    return out_path


def _quick_diff(rendered: Path, golden: Path) -> str:
    if not golden.exists():
        return "no-golden"
    a = rendered.stat().st_size
    b = golden.stat().st_size
    drift = abs(a - b) / max(a, 1)
    return "drift" if drift > 0.02 else "ok"


def _parse_args(argv: list[str]) -> argparse.Namespace:
    p = argparse.ArgumentParser(description="Render gm-console Web SPA screenshots.")
    p.add_argument("--base-url", default="http://127.0.0.1:8080")
    p.add_argument("--out-dir", default="apps/gm-console-web/dist/screenshots")
    p.add_argument("--device-filter", default=None, help="comma-separated device ids")
    p.add_argument("--update-goldens", action="store_true")
    return p.parse_args(argv)


def main(argv: list[str] | None = None) -> int:
    args = _parse_args(list(argv) if argv is not None else sys.argv[1:])
    out_dir = Path(args.out_dir).resolve()
    devices: Iterable[str] | None = (
        [d.strip() for d in args.device_filter.split(",") if d.strip()] if args.device_filter else None
    )
    shots = [s for s in SHOTS if devices is None or s.device_id in devices]
    driver = _load_driver()

    any_drift = False
    for shot in shots:
        rendered = _render_one(driver, args.base_url, out_dir, shot, args.update_goldens)
        golden = out_dir / "goldens" / shot.filename
        status = _quick_diff(rendered, golden)
        print(f"[render] {shot.shot_id} @ {shot.device_id} ({shot.width}x{shot.height}) -> {rendered} ({status})")
        if status == "drift":
            any_drift = True
    return 4 if any_drift else 0


if __name__ == "__main__":
    raise SystemExit(main())